src/handler/core.rs

```diff
@@ -44,36 +52,71 @@ use crate::{description, HandlerDescription};
 /// This is very crucial when `b` is a filter: if it is chained, it decides
 /// whether or not to call `c`, but when it is branched, whether `c` is called
 /// depends solely on `a`.
-pub struct Handler<'a, Input, Output, Descr = description::Unspecified> {
-    data: Arc<HandlerData<Descr, DynF<'a, Input, Output>>>,
+pub struct Handler<'a, Output, Descr = description::Unspecified> {
+    data: Arc<HandlerData<Descr, DynF<'a, Output>>>,
 }
 
 struct HandlerData<Descr, F: ?Sized> {
     description: Descr,
+    sig: HandlerSignature,
     f: F,
 }
 
-type DynF<'a, Input, Output> =
-    dyn Fn(Input, Cont<'a, Input, Output>) -> HandlerResult<'a, Input, Output> + Send + Sync + 'a;
+/// A run-time handler signature. Used only for run-time type inference and
+/// checking of handler chains.
+///
+/// See [`crate::type_check`].
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HandlerSignature {
+    /// The signature corresponding to [`crate::entry`].
+    Entry,
```

Comment:
why is this needed at all? isn't `entry()`, for purposes of type checking DI just a `input_types = [], obligations = {}`?

Reply:
Then we would need `output_types` to be the set of all types, because otherwise, an intersection of any sequence of branches with an entry branch would be always empty.